pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -412,7 +412,7 @@ func (t *ResolvedPipelineTask) skipBecauseResultReferencesAreMissing(facts *Pipe
 		resolvedResultRefs, pt, err := ResolveResultRefs(facts.State, PipelineRunState{t})
 		rpt := facts.State.ToMap()[pt]
 		if rpt != nil {
-			if err != nil && (t.IsFinalTask(facts) || rpt.Skip(facts).SkippingReason == v1beta1.WhenExpressionsSkip) {
+			if err != nil && (t.IsFinalTask(facts) || rpt.skipBecauseWhenExpressionsEvaluatedToFalse(facts)) {
```

Comment:
Is `skipBecauseWhenExpressionsEvaluatedToFalse` direct substitution of  `rpt.Skip(facts).SkippingReason == v1beta1.WhenExpressionsSkip`?

In this function `skipBecauseResultReferencesAreMissing`, the task results in `finally` tasks or `whenExpressions` in `dag` tasks, or `whenExpressions` in `finally` tasks  are resolved.

If the result resolution fails and returns an error, then it checks if it's a `finally` task or if this task appears in the list of  `Skipped` tasks with `skipping` reason set to `WhenExpressionsSkip`.

`skipBecauseWhenExpressionsEvaluatedToFalse` just checks if the `WhenExpression` evaluates to `true` or `false` irrespective of if it’s cached in the list of skipped tasks.

Do we need to maintain such `cache` if it's not useful? The `cache` was introduced for huge pipelines with large set of dependencies and it sure improved the task scheduling.

Reply:
Also, the cache is still very much useful, it's just that something (and again, I haven't been able to figure out what exactly) is happening when you switch `enable-api-fields` to `alpha` mid-`PipelineRun` that's resulting in infinite recursion. It's definitely an edge case.

Maybe the answer here is actually just to flip the order of the `case`s for `t.skipBecauseResultReferencesAreMissing(facts)` and `t.skipBecauseWhenExpressionsEvaluatedToFalse(facts)` in `skip`? Hmm, probably not, because you need to be sure you've got all results that could be referenced in when expressions before evaluating the when expressions...this one's a bit ugly.